### ISSUE TITLE
Support IPv6

### ISSUE THREAD

**Issue by WilliamDEdwards** [moby/moby#44923]
### Description

Maybe Docker/Moby should start supporting IPv6.

And by 'supporting' I don't mean having to skim through GitHub comments and the Docker forum for hours to enable basic functionality.

I tried to enable IPv6 for user-defined bridge networks.

1.

To enable user-defined bridge networks, `default-address-pools` must be set. I set `default-address-pools`, causing Docker to not start anymore.

There's [this months old issue](https://github.com/moby/moby/issues/44220#issuecomment-1379529812) that says an IPv4 address address pool should be before placed before an IPv6 address pool.

Aside from the code smell that this bug brings, this requires me to set an IPv4 address pool. Why?

2.

I attempted to remove `fixed-cidr-v6`, as I will only be using user-created networks, which caused Docker not to start anymore.

I found [this forum post](https://forums.docker.com/t/assigning-default-ipv6-addresses/128665), which says:

> The subnet specified in fixed-cidr-v6 defines what the docker0 network on the machine should look like. No more, no less. I still think this is poorly documented.

And indeed, the documentation says that IPv6 cannot be disabled for the default bridge. However, nowhere does the [documentation](https://docs.docker.com/config/daemon/ipv6/) say that `fixed-cidr-v6` is the subnet used for the `docker0` bridge.

3.

To add the IPv4 address pool (which should not be needed), I need to know what the IPv4 address pool is.

This is not documented anywhere. I found [this almost 4 year old issue](https://github.com/docker/docs/issues/8663#issuecomment-956438889) requesting for this to be documented.

4.

Finally having gotten Docker to stop complaining about missing or misordered settings, Docker crashed immediately on startup because it used all available RAM.

I found [this almost 4 year old issue](https://github.com/moby/moby/issues/40275), which says that at least an /80 should be used. Makes sense, but why does Docker crash because of this 4! years after the bug report?!

**Comment by neersighted** [comment]
I'm not sure this is a productive format to try to roll up known issues -- but @thaJeztah has a much better idea of what tracking issues we already have and can probably chime in. If we don't have a good tracking issue, this could be it (assuming that we edit the body to be readable).

@corhere is in the process of working on a couple issues in libnetwork, namely the memory issue (the daemon will attempt to represent every address in memory, which as you can imagine, rapidly becomes impossible). It sounds like the `fixed-cidr-v6` behavior might also be worth investigating.

**Comment by polarathene** [comment]
## TL;DR

All concerns raised in the issue are effectively resolved or in progress to be.

Docs however could improve for IPv6 (_but could be tracked elsewhere_):
- `default-address-pools` (_should be sufficient to just provide private IPv4 + IPv6 subnets example_)
- Common ULA config example
- Optional GUA example + notes/gotchas

## Summary

1. **User-defined bridge networks don't require address pools configured** to be created. You can set subnets explicitly instead. The IPv4 before IPv6 ordering requirement has addressed for v24 release IIRC.
2. `fixed-cidr-v6` is only for the default bridge, it's not a helpful failure but shouldn't be occur if the config disables the `ipv6` setting (_also only relevant to default bridge_). The **default bridge is considered legacy** (_this is documented officially too_), it is just convenient for implicit network config.
3. Default address pools are not documented, but **`daemon.json` docs have an example that should be adequate**. It would be better to see `default-address-pools` documented better with the two IPv4 pools that defaults usually cover though. Less user error then.
4. The memory failure is being resolved. You can **avoid it by creating an IPv6 network with an explicit subnet** instead of using address pools (_or don't create pools with billions of subnets_). Subnets don't need to be `/80` in size either, the concern no longer seems applicable as 48-bit MAC address of container is not truncated in smaller subnets, address changes to incremental / predictable assignment.

---

- Most issues are niche, a tracking issue for IPv6 doesn't seem necessary now that there is a tag being assigned specifically for IPv6.
- [IPv6 Docs](https://docs.docker.com/config/daemon/ipv6/) could probably be improved:
  - Presently only shows `ipv6` + `fixed-cidr-v6` settings to enable on default bridge for a `/64`. Neither setting is required for IPv6 user-defined networks, which also do not require default address pools configured as OP was given the impression (_would be simpler to understand once the memory bug is fixed_).
  - Documentation IP used, but users may not be familiar with ULA and GUA addresses or appropriate subnets:
    - ULA would be most appropriate and easiest/consistent to configure, but should mention `ip6tables`.
    - GUA may often be in environments where the docker bridge network can't use a `/64` and needs to handle proxying (NDP) + environment specific config that affect connection being reachable or reliable.

## Detailed responses

> And by 'supporting' I don't mean having to skim through GitHub comments and the Docker forum for hours to enable basic functionality.

> If we don't have a good tracking issue, this could be it (assuming that we edit the body to be readable).

It'd probably be helpful to have, but the bigger issue is probably lack of good docs for the common IPv6 setup.
- The docs used to cover much more in detail [until a PR dropped them without much discussion](https://github.com/moby/moby/issues/40275#issuecomment-1505068182).
- [Another PR was opened to improve it](https://github.com/docker/docs/pull/8292) but went stale from lack of additional review, the Docker bot closed it.

I've not had much IPv6 knowledge myself until recently, users were having issues with a Docker container I help maintain not being able to resolve IPv6 remote client address, getting a docker network gateway IPv4 or IPv6 address instead.

<details>
<summary>Expand for more details</summary>

- `ip6tables` with config for user-defined network or default bridge provides the first requirement if wanting a familiar IPv4 NAT experience with port publishing.
- Docker compose config has some config to be aware of when relevant, and one needs to be aware of implicit vs explicit subnet assignment when creating networks.
- Private IPv6 with ULA works well, but similar to IPv4 private range, users may not be knowledgeable enough to know how to configure that when a default address pool is lacking. Examples in docs would probably assist that.
  - Slight gotcha that can be encountered with DNS resolution IPv4 private address can be resolved over IPv6 private (ULA) IIRC, usually due to config such as `/etc/gai.conf` in Docker host environment, out of it's control. With GUA, IPv6 will have precedence, when not being explicit about protocol, this can some times be confusing to encounter and troubleshoot/understand.
- Public IPv6 with GUA is a bit more tricky. `ip6tables` is not required as this differs by not needing to publish container ports. Firewall rules (at least with UFW) aren't bypassed when NAT rules aren't added from Docker, and if firewall is present needs more explicit configuration, or risk of exposing all ports unintentionally to be publicly reachable for that IPv6 GUA network/address. The other common pain point that seemed to trip up users more was for non-routed IPv6 `/64` blocks. Docker host creating a subnet smaller than `/64` from that external interface via it's bridges requires kernel config or the external interface to be adjusted.
  - `proxy_ndp` enables the container to be reachable from remote clients (_but can be reported as unstable / spotty, likely due to interface being reset/recreated from event triggers, as I experienced for `cloud-init` VPS_).
  - Manually adding each container IP as a proxy address for the external interface to respond to (_flushed between restarts_). This can be automated via several daemon services, or Docker could implement the support (_some issues request, a PR provided an implementation, there's other Go based implementations too_).
  - `accept_ra=2` setting is apparently required. There's an issue about Docker setting `forwarding=1` affecting this setting elsewhere and breaking connectivity in some configurations. `accept_ra=0` is also set by systemd `networkd` which implements it's own implementation internally, thus can be a bit misleading if the user is not aware and trying to follow some guide without configuring `networkd` explicitly (_eg: `cloud-init` config from VPS vendor_).

</details>

---

> 1. To enable user-defined bridge networks, `default-address-pools` must be set.
>      I set `default-address-pools`, causing Docker to not start anymore.

This is not true. You can create your own networks with a specific subnet instead of assigning one implicitly from the default address pools:

```bash
docker network create \
  --ipv6 \
  --subnet fd00:cafe:feed:face:f00d::/80 \
  --opt com.docker.network.bridge.name=br-example \
  example-ipv6
```

The `--opt` is not necessary, but helps identify the bridge interface outside of Docker (eg: `ip a`), useful if you have to reference it for firewall rules (_such as for IPv6 GUA address, unlike the ULA address example above_).

<details>
<summary>Remainder of 1 explained (click to view)</summary>

> There's [this months old issue](https://github.com/moby/moby/issues/44220#issuecomment-1379529812) that says an IPv4 address address pool should be before placed before an IPv6 address pool.
> 
> Aside from the code smell that this bug brings, this requires me to set an IPv4 address pool. Why?

The IPv4 requirement is from the network stack being built around IPv4, while IPv6 support was added much later AFAIK. They've been refactoring it a fair amount and will eventually reach that goal of being able to support IPv6 only (_although I don't think that is practical for deployments with public clients, my personal system can't reach hosts via IPv6 for example_).

The referenced issue looks like it got addressed and shouldn't be an issue in a future release that includes the fix?

</details>

---

> 2. I attempted to remove `fixed-cidr-v6`, as I will only be using user-created networks, **which caused Docker not to start anymore**.

I don't think this is exactly true. You need to first disable `ipv6` again (_only relevant for the default bridge_).

<details>
<summary>Explanation (click to view)</summary>

I have experienced it when adjusting the settings in `/etc/docker/daemon.json`. There was definitely a way to reproduce the bug where `dockerd` would not start due to a change related to this setting (_A recent one was also introduced requiring `bip` to be set but is unrelated_).

It was fixable, but required other settings to be adjusted first. IIRC when `dockerd` / `containerd` starts, it does some cleanup on networks that aren't relevant to the new config. After that is handled, a setting might not be to toggled before you can remove `fixed-cidr-v6`... something like that :sweat_smile: (_I believe it was requiring `ipv6` setting to be disabled first, restart `dockerd`, then you can change / remove `fixed-cidr-v6` and restart again_)

</details>

> And indeed, the documentation says that IPv6 cannot be disabled for the default bridge. However, nowhere does the [documentation](https://docs.docker.com/config/daemon/ipv6/) say that `fixed-cidr-v6` is the subnet used for the `docker0` bridge.

Regarding that IPv6 statement, that's a fault of the docs being a bit incorrect (_at least now_). There is some docs that relate `fixed-cidr-v6` to the default bridge, but it's could be communicated better for sure.

<details>
<summary>Explanation (click to view)</summary>

It could be better explained in the [reference `daemon.json` / `dockerd` CLI options page](https://docs.docker.com/engine/reference/commandline/dockerd/) (_which I've found frustrating UX to reach as there is no link from the config docs sections related to it_).

If you visit the [networking bridge docs](https://docs.docker.com/network/bridge/#configure-the-default-bridge-network), you'll see at the bottom it does identify a group of related settings, which should help establish the relevance of `fixed-cidr-v6` from that context though. It definitely wasn't clear to me either when I first started looking into such config, and had to mostly learn from trial and error, until I'd recognize snippets scattered in docs / issues as relevant to.

[Those docs also describe IPv6 on the default bridge](https://docs.docker.com/network/bridge/#use-ipv6-with-the-default-bridge-network) cannot be selectively disabled. There's also an earlier section further up on that page mentioning similar, but in both cases it seems a bit misleading:
- AFAIK, default bridge `docker0` or your own user-defined bridge networks can enable IPv6 and containers will be assigned IPv6 addresses, neither has an opt-out for containers on the network? So it's unclear what it's trying to communicate there.
- `ipv6` does not need to be enabled in `daemon.json` for user-defined networks to use IPv6, regardless of that being implied via the current docs. This setting only affects enabling IPv6 on the default bridge, which only matters if you don't specify a network for the container to be joined to (_unless using compose, which creates it's own default bridge network by default unrelated to `docker0` settings_).

</details>

---

> 3. To add the IPv4 address pool (which should not be needed), I need to know what the IPv4 address pool is.
> 
> This is not documented anywhere. I found [this almost 4 year old issue](https://github.com/docker/docs/issues/8663#issuecomment-956438889) requesting for this to be documented.

This is still a problem and it's a bit sad that the automation is closing some valid issues or PRs (that had partial approval) but didn't get resolved.

Realistically, you only need to configure an IPv4 private address pool with that can supply enough networks implicitly. You don't actually need to match the undocumented defaults, but I don't expect many users to understand how to configure the pools properly (_the bigger documentation issue_).

An example is documented which should be sufficient to reference blindly, but you need to know where to find it in the docs, and understand that it's required/applicable.

<details>
<summary>Explanation (click to view)</summary>

I had first come across [this resource](https://straz.to/2021-09-08-docker-address-pools/) explaining it well. I don't think it's important to document the implicit defaults too much as you likely don't need to reproduce them when changing the default pools.

The [`daemon.json` reference docs](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file) does have an example with two IPv4 pools of `/16` split into `/24` subnets, that is 256 networks each with 256 IPv4 addresses (`2^(32-16) / 2^(32-24) == 256`). You're not likely to need more than that. Or you can follow the linked blog resource I provided:

```json
{
  "default-address-pools" : [
    {
      "base" : "172.17.0.0/12",
      "size" : 16
    },
    {
      "base" : "192.168.0.0/16",
      "size" : 20
    }
  ]
}
```

Roughly matches the Docker default pools IIRC. That's 31 networks each with 4096 (`2^(32-20)`) or approx 65k (`2^(32 - 16)`) IP addresses (_Technically a little bit less due to how subnets require reserving a few addresses to work_). More than you'll need :+1: 

This will be less of an issue if a future release later decides to add ULA subnet into the default address pool as well for IPv6 networking with NAT for an experience similar to networking with IPv4. That requires the `ip6tables` setting to come out of `experimental` and be enabled though I think?

The address pools are convenient, but I'm not sure how important they are. They seem to mostly simplify network creation. One important part that can be error prone is if the user does not know about private IP address ranges. I have seen some assume `172.x.y.z` is usable, but `x > 31` is outside of the private range IIRC. So defaults are helpful when you don't have to understand how to choose safe subnets :sweat_smile: (_I made that mistake years ago myself_)

</details>

---

> 4. Docker crashed immediately on startup because it used all available RAM.

If you don't try to configure such a large address space, that shouldn't be a problem.

The memory failure aside, this is the result of users not understanding what they're configuring (_due to poor documentation of configuring pools_). It's applicable to IPv4 for similar reasons, and is only a problem with IPv6 pools due to how many subnets are permitted.

<details>
<summary>Explanation (click to view)</summary>

As mentioned, that area of code is being reworked to be a bit smarter since the address space with IPv6 is massive. An `/80` subnet is 48-bit in size (281 trillion?) and your only likely to have less than 100 containers on that network right?

`/112` would be 16-bit subnet size (approx 65k IP addresses), and is more inline with what you'd see on IPv4 subnets. That said the `size` setting for those pools, splits up the subnet range into smaller subnets. IIRC that was the actual issue, creating a pool of a very large set of subnets, not the size of subnets themselves (_that is 8 subnets with billions of IPs is not a problem, but billions of subnets with not as many IP is problematic_).

Anyway, not relying on implicit address pool and just providing a `--subnet` value avoids that concern entirely, you'll just get what you asked for and it doesn't seem to cause any memory issues in my experience when you do it this way :)

</details>

> which says that at least an `/80` should be used, makes sense

This isn't really relevant anymore. You should be able to use smaller or larger subnets without issue.

<details>
<summary>Explanation (click to view)</summary>

- If you have a subnet larger than `/80` (eg: `/79`) on the default bridge it'll still use the 48-bit MAC address AFAIK, and presumably limited to `/80` in range (_not that you'd exceed that on a single network_).
- If you have a smaller subnet (eg: `/81`) on the default bridge, it'll no longer use the 48-bit MAC address and be a more predictable container address that is assigned incrementally like with IPv4. So no NDP cache issue should occur AFAIK?
- User-defined networks AFAIK do not mimic this `/80` or larger subnet behaviour with MAC address of a container defining it's IP (_none of this is officially documented though IIRC, which doesn't help_).

I will admit the advice was originally not easy to understand when I first encountered it. I wasn't sure if it was Docker specific or outdated.

It turned out to be [from old Docker docs but PRs for it via `git blame` didn't reveal much](https://github.com/moby/moby/issues/40275#issuecomment-1505068182). Back then I also didn't understand if _"at least `/80`"_  meant `/79` or `/81` was when things would go bad, or if it was something else entirely. Especially since typically you'd see advice that IPv6 subnets should be no smaller than `/64`, but that seems to have some overlapping terminology where it's a non-issue, and more to do with "routed" subnets or CIDR blocks I think.

</details>

**Comment by mhoam** [comment]
Many thanks for this explanation of the current situation. I have been trying for weeks to get IPv6 working on a PiHole container, built using docker compose.
As an amateur newbie, I found the documentation on IPv6 difficult and confusing so encountered many of the misunderstandings, errors and confusion described.
I now have a much better grasp on how to proceed 
:-)

**Comment by polarathene** [comment]
@mhoam oh that's unfortunate as the Docker IPv6 docs were overhauled recently. Was there anything in particular that they could improve on?

I wrote similar IPv6 docs for a docker-focused project which might be slightly better: https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/ipv6/#enable-proper-ipv6-support

Not sure if those are also lacking some information that is better covered here.

**Comment by mhoam** [comment]
@polarathene 
Thanks for your "Detailed Response" above and link to the very useful document. IMHO your document is much more useful to me as it shows the use of IPv6 in a Compose scenario. My requirement is simple, one container running PiHole in Docker created by Compose to use IPv6 alongside IPv4.
I encountered much the same problems as @WilliamDEdwards. i.e when reading the document https://docs.docker.com/config/daemon/ipv6/ it just wasn't clear re the distinction between the use of the daemon for the default network and/or what is required before using and within the yml of a Compose created network.
I also seemed to have an issue when changing the daemon.json, some times it needed just a restart, other times it required a reboot.
I also hit a "system-limit-hit" error when trying to run the Compose file. No indication as to which limit or why that had occurred! Searching produced no useable info so I guess that is a recent error trap? I now understand that was probably caused by Compose trying to build an enormous IPv6 network, not so clear when it failed after hours of trying and failing :-(

There is lots of info posted elsewhere by those trying to get IPv6 working in Docker over the last 6 or 7 years. Some of that was unhelpful and some actually misleading now. Of course, nothing can be done about that, the starting point for most should be the "Enable Ipv6 support" Docker page so perhaps that should make it clear that the level of IPv6 support has changed and is still changing?

Thanks again
Ron

**Comment by polarathene** [comment]
Cheers for the feedback! 😁 

> IMHO your document is much more useful to me as it shows the use of IPv6 in a Compose scenario.

I had a quick look, but `compose.yaml` wise there's only a small snippet to show IPv6 config there, that the official Docker IPv6 docs have too. The rest is generic network config that applies for standard IPv4 too.

`docker-mailserver` IPv6 docs:
<img width="422" alt="image" src="https://github.com/moby/moby/assets/5098581/f31c7dd0-6464-4e2a-8f5e-6ccea9b6c8c3">

Official Docker IPv6 docs:
<img width="394" alt="image" src="https://github.com/moby/moby/assets/5098581/8f78a523-5fd0-409d-8523-055dc827b66d">

> it just wasn't clear re the distinction between the use of the daemon for the default network and/or what is required before using and within the yml of a Compose created network.

I see, I think I may have raised a similar concern during the review process. You just need to cover the initial `daemon.json` config and restart the daemon, then you can create IPv6 networks like the previous screenshot showed:

<img width="400" alt="image" src="https://github.com/moby/moby/assets/5098581/7f2ab5d1-0755-4b4f-818d-ae3ffa01a3b3">

While the next section of their docs shows the default bridge support, which is just adding two more settings to `daemon.json` to enable IPv6, and assign an IPv6 network that IP addresses can be provided from (_they use a reserved for documentation IPv6 subnet as an example_), no network needs to be created since it just adds IPv6 to the default bridge which containers use implicitly:

<img width="401" alt="image" src="https://github.com/moby/moby/assets/5098581/c69cd1d4-787e-4e90-b180-cf0b26350646">

The next section might need more clarification (_at least within the context of the prior default bridge section_). This is only to add IPv6 networks to the default pools for Docker to manage IPv6 networks in the same way it does implicitly for IPv4, not requiring you to configure as much:

<img width="386" alt="image" src="https://github.com/moby/moby/assets/5098581/30e9eaca-05ac-4c85-9067-7bb1e071c617">

You can just follow the first section in those docs (_and choose an IPv6 ULA subnet like `fd00:cafe:face:feed::/64`_) and ignore the other two sections. Just assign the IPv6 network explicitly and you minimize the problems from behind-the-scenes magic 😝 

---

> I also seemed to have an issue when changing the `daemon.json`, sometimes it needed just a restart, other times it required a reboot.
> 
> I also hit a "system-limit-hit" error when trying to run the Compose file. No indication as to which limit or why that had occurred! Searching produced no useable info so I guess that is a recent error trap? I now understand that was probably caused by Compose trying to build an enormous IPv6 network, not so clear when it failed after hours of trying and failing

I think that might be as you suggest with the network size, you'll find [common advice online suggesting large networks like `/80`](https://github.com/moby/moby/issues/40275#issuecomment-1505068182) (_48-bit wide range of addresses, over 280 trillion IPv6 addresses_), but this doesn't cause the problem.

It's the number of networks if you configure `default-address-pools` in the `daemon.json` to take a supernet `base` of `/64` 
and split that into smaller `size` subnets like `/80`, this would create 65k docker IPv6 networks. Reducing the `size` (_a larger number_) of the subnet pool without adjusting the `base` worsens that outcome, as you can subdivide the `base` into many smaller subnets (_a `size` of `/96` would split into over 4 billion networks for a `base` of `/64`_ 😅 )

That's more of the fault of Docker being a bit too eager and doing all that extra work (allocations) when it doesn't need to in advance since you're not actually using all of those. There is WIP to resolve that by [creating the networks lazily](https://github.com/moby/moby/pull/43033) (on-demand).

**In the meantime**, you can still configure the  `default-address-pools` - just choose a number like subnet pool `size` of `/80` or larger (_a smaller number_). You can also decrease the size of the `base` (supernet?) range that gets split up into smaller networks. On the [official IPv6 Docker docs](https://docs.docker.com/config/daemon/ipv6/#dynamic-ipv6-subnet-allocation) this is shown with [`base`: `/104` and `size`: `/112`](https://github.com/docker/docs/pull/17176#discussion_r1212571285) (_I suggested this information during review_), and as the note in this section sates, the [difference between `base` and `size` should really try not to exceed `24`](https://github.com/docker/docs/pull/17176#discussion_r1212549431).

Or configure a single `/64` IPV6 ULA network explicitly instead like my docs show (_containers will get IPv6 assigned from that network_). I just noticed that I didn't touch on configuring `default-address-pools` much for `docker-mailserver` IPv6 docs, probably because it'd risk confusing our users with these kinds of issues.

---

> There is lots of info posted elsewhere by those trying to get IPv6 working in Docker over the last 6 or 7 years. Some of that was unhelpful and some actually misleading now
> 
> so perhaps that should make it clear that the level of IPv6 support has changed and is still changing?

Yeah, it was very confusing for me too as a newbie to IPv6 and how to set that up with Docker (_before May there was much less information in the official Docker docs_), they updated theirs roughly around the same time that I was making a big update to the IPv6 docs from the project I work on.

Support for IPv6 in Docker has been improving over the years, but more recently I think it (_and networking in general for Docker_) has been getting some much-needed love :)

**Comment by mhoam** [comment]
@polarathene Thanks for the detailed response.
That snippet really clarified my thinking:-)
Not much time today but I will be trying again this week and will provide an update as and when,
Cheers
Ron

**Comment by mhoam** [comment]
@polarathene 
The IPv6 setup as described in your MailServer document does NOT work for me :-)
When I run the Compose file I receive an error "networks.pihole6 Additional property subnet is not allowed" It appears I cannot configure the IPv6 subnet to be used at that point.
My Compose file is running at version 3.9 and the "additional Properties" section seems to have caused many different issues previously. Can I ask what version of Compose was used in your mailserver IPv6 example?
The Docker Compose docs highlight the change from V1 to V2 Compose and the latest version is V2.21 (https://docs.docker.com/compose/release-notes/) 
The starting point for my compose.yml was an example on the pihole/pihole Docker page, not sure why it references "Version 3.9"
Looking through the Release Notes I can only find one reference to "subnet" which was adding validation

This page https://docs.docker.com/compose/compose-file/05-services/#ipv4_address-ipv6_address
appears to suggest that an "ipam" section is needed to specify a fixed IP address, either IPv4 or IPv6. I'm not sure what ipam actually is!

Any suggestions?

**Comment by polarathene** [comment]
> My Compose file is running at version 3.9 and the "additional Properties" section seems to have caused many different issues previously. Can I ask what version of Compose was used in your mailserver IPv6 example?

We are aligned to upstream Docker docs for Compose, which since June 2023 if I remember correctly, switched over to V2 and `docker-compose.yml` => `compose.yaml` (_although this is not required_).

> not sure why it references "Version 3.9"

`version` is also not required in the compose config with Compose V2. It enables support of some Compose V1 Schema V2 + V3 settings which the latest Compose V1 had dropped. This simplified network syntax might be specific to Compose V2 though.

I'd suggest installing a newer version of Compose if you can. The official Docker docs has guides for a variety of distros.

---

> appears to suggest that an "ipam" section is needed to specify a fixed IP address, either IPv4 or IPv6. I'm not sure what ipam actually is!

IPAM => Internet Protocol Address Management ([I think](https://github.com/docker-mailserver/docker-mailserver/pull/3244#issuecomment-1624353218)?)

That was [confusing to me when I first encountered it too](https://github.com/docker-mailserver/docker-mailserver/pull/3244#issuecomment-1620757960), and I guess the improvement (_in Compose V2?_) to not require it, and hoist these settings up implicitly might have been for UX reasons 🤷‍♂️ 

I [originally had advice for the config with `ipam`](https://github.com/docker-mailserver/docker-mailserver/pull/3244#issuecomment-1623406705) while working on those IPv6 docs, and was asked about the difference during feedback here:
https://github.com/docker-mailserver/docker-mailserver/pull/3244#issuecomment-1617880777

For consistency with the updated official Docker IPv6 docs in May, we also went with the simplified config without the nested `ipam` section.

---

> Any suggestions?

Get Compose V2:
- Via `docker` plugin: https://docs.docker.com/compose/install/linux/#install-using-the-repository
- Standalone `docker-compose` binary: https://docs.docker.com/compose/install/standalone/

You may also want to get a newer version of Docker Engine if you're not on v23 or newer (_released early this year_): https://docs.docker.com/engine/install/raspberry-pi-os/

Alternatively, use the `ipam` section, as I originally documented in a comment before writing nicer docs: https://github.com/nginx-proxy/nginx-proxy/issues/133#issuecomment-1368745843

**Comment by mhoam** [comment]
Thanks @polarathene 
My 32bit RaspiOS is currently running Docker Engine (client and server) at version 24.0.5. Compose is at version V2.20.2 

Sadly I still get "additional property subnet is not allowed" when trying to start the container :-(

**Comment by polarathene** [comment]
> Sadly I still get "additional property subnet is not allowed" when trying to start the container :-(

Even when you try to use it with `ipam`?

```yml
networks:
  # Overrides the `default` compose generated network, avoids needing to attach to each service:
  default:
    enable_ipv6: true
    # An IPv4 subnet is implicitly configured, IPv6 needs to be specified:
    ipam:
      config:
        - subnet: fd00:cafe:face:feed::/64
```

**Comment by mhoam** [comment]
Using the "ipam" syntax the container starts and seems to have the correct IPv6 address. I need to test it further to make sure Pihole is working correctly on IPv6 DNS requests.
However for the sake of this discussion, the "ipam" syntax works to define the IPv6 address to be used within the container BUT the simpler "subnet" syntax as described in your mailserver doc does not work, at least in my environment. 

Let me know if you would like me to test another approach to configuring an IPv6 ULA for a container.
I will update you again when I have tested the use of IPv6 addresses into and out of that container
Thanks for your help 
Ron

**Comment by polarathene** [comment]
I'm really not sure where to go from there.

Perhaps raise a bug report [here](https://github.com/docker/compose) about it (_with as much of the `compose.yaml` content that you can share / anonymize_), The maintainers there might know what's going wrong 🙏

**Comment by schildbach** [comment]
Wouldn't it be low-hanging fruit to simply add proper (ULA, not "for documentation") IPv6 address pools to the default address pools hardcoded in dockerd? It would make IPv6 configuration already a lot simpler. It seems so obvious and I wonder why it hasn't been done already?
